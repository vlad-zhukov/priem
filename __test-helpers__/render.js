/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import {render} from 'react-testing-library';

export default (element, options) => {
    const ref = React.createRef();
    const el = React.cloneElement(element, {ref});
    const result = render(el, options);
    result.instance = ref.current;
    return result;
};
