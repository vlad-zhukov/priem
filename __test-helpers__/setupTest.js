/* eslint-disable import/no-extraneous-dependencies */

import 'raf/polyfill';
