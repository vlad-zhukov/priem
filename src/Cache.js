import moize from 'moize';
import {elementsEqual} from 'react-shallow-equal';
import * as promiseState from './promiseState';
import {type, assertType, isBrowser} from './helpers';

function arrayElementsEqual(a, b) {
    if (!Array.isArray(a) || !Array.isArray(b) || a.length !== b.length) {
        return false;
    }
    for (let i = 0; i < a.length; i++) {
        if (a[i] !== b[i]) {
            return false;
        }
    }

    return true;
}

export const defaultAsyncValueOptions = {autoRefresh: true, maxSize: Infinity};

export default class Cache {
    constructor() {
        this.memoized = null;
        this.awaiting = [];
        this.rejected = false;
        this.prevArgs = null;
    }

    add(value, onExpire) {
        assertType(value, ['object'], "'getAsyncValue() -> value'");
        assertType(value.promise, ['function'], "'getAsyncValue() -> value.promise'");

        this.memoized = moize(value.promise, {
            isPromise: true,
            maxAge: value.maxAge,
            maxArgs: value.maxArgs,
            maxSize: value.maxSize,
            onExpire,
        });
    }

    isMemoized(args) {
        if (!this.memoized) {
            return false;
        }

        return this.memoized.has(args);
    }

    getAwaitingIndex(args) {
        for (let i = 0, l = this.awaiting.length; i < l; i++) {
            if (elementsEqual(this.awaiting[i], args)) {
                return i;
            }
        }

        return -1;
    }

    isAwaiting(args) {
        return this.getAwaitingIndex(args) !== -1;
    }

    removeAwaiting(args) {
        const index = this.getAwaitingIndex(args);
        if (index !== -1) {
            this.awaiting.splice(index, 1);
        }
    }

    run({asyncValue, isForced, update, onExpire}) {
        const value = {...defaultAsyncValueOptions, ...asyncValue};

        // Stop auto-refreshing
        if (!value.autoRefresh && !isForced) {
            return;
        }

        const args = type(value.args) === 'array' ? value.args : [];

        if (!this.memoized) {
            this.add(value, onExpire);
        }

        // Do not recall rejected (uncached) promises unless forced
        if (this.rejected && !isForced) {
            return;
        }

        if (this.isAwaiting(args)) {
            return;
        }

        let shouldUpdateState = true;

        // eslint-disable-next-line consistent-return
        update((s, m) => {
            const isFulfilled = !!s.fulfilled;
            const isSsr = !!(isFulfilled && m.ssr);

            if (this.isMemoized(args)) {
                // Do not recall memoized promises unless forced
                if (isFulfilled && !isForced && arrayElementsEqual(args, this.prevArgs)) {
                    shouldUpdateState = false;
                    return null;
                }
            }
            else if (isSsr) {
                shouldUpdateState = false;
                this.memoized.add(args, Promise.resolve(s.value));
                this.prevArgs = args;
                // eslint-disable-next-line consistent-return
                return {
                    meta: {ssr: false},
                };
            }

            this.awaiting.unshift(args);
            this.rejected = false;

            if (isFulfilled) {
                return {
                    state: promiseState.refreshing(s),
                };
            }

            return {
                state: promiseState.pending(),
            };
        });

        if (!shouldUpdateState) {
            return;
        }

        if (isForced) {
            this.memoized.remove(args);
        }

        // eslint-disable-next-line consistent-return
        return this.memoized(...args)
            .then((result) => {
                this.removeAwaiting(args);
                this.prevArgs = args;
                update({
                    state: promiseState.fulfilled(result),
                    meta: {ssr: !isBrowser},
                });
            })
            .catch((e) => {
                this.removeAwaiting(args);
                this.rejected = true;
                this.prevArgs = null;
                update({
                    state: promiseState.rejected(e.message),
                    meta: {ssr: !isBrowser},
                });
            });
    }
}
