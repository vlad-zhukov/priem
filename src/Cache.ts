const NEXT = '@@next';
const PREV = '@@prev';

export class CacheItem<K, V> {
    key: K;
    value: V;
    [NEXT]: CacheItem<K, V> | undefined;
    [PREV]: CacheItem<K, V> | undefined;
    isValid!: boolean;
    expireTimerId: number | undefined;
    lastUpdateAt: number | undefined;

    constructor(key: K, value: V) {
        this.key = key;
        this.value = value;

        Object.defineProperties(this, {
            [NEXT]: {
                value: undefined,
                writable: true,
            },
            [PREV]: {
                value: undefined,
                writable: true,
            },
            isValid: {
                value: false,
                writable: true,
            },
            expireTimerId: {
                value: undefined,
                writable: true,
            },
            lastUpdateAt: {
                value: undefined,
                writable: true,
            },
        });
    }

    destroy(): void {
        delete this.key;
        delete this.value;
        this[NEXT] = undefined;
        this[PREV] = undefined;
        this.isValid = false;
        clearTimeout(this.expireTimerId);
        this.expireTimerId = undefined;
        this.lastUpdateAt = undefined;
    }
}

export interface SerializableCacheItem<K, V> {
    key: K;
    value: V;
}

const REDUCED = '@@reduced';

interface ReducedType<ValueType> {
    [REDUCED]: true;
    value: ValueType;
}

function reduced<ValueType>(value: ValueType): ReducedType<ValueType> {
    return {[REDUCED]: true, value};
}

function isReduced<ValueType>(value: any): value is ReducedType<ValueType> {
    return value && REDUCED in value;
}

export function reduce<ReturnType, K, V>(
    cache: Cache<K, V>,
    accumulator: ReturnType,
    iteratee: (result: ReturnType, item: CacheItem<K, V>) => ReturnType | ReducedType<ReturnType>,
) {
    let item = cache.head;
    let result: ReturnType | ReducedType<ReturnType> = accumulator;
    while (item) {
        result = iteratee(result, item);
        if (isReduced(result)) {
            return result.value;
        }
        item = item[NEXT];
    }
    return result;
}

export class Cache<K, V> {
    head: CacheItem<K, V> | undefined;
    tail: CacheItem<K, V> | undefined;
    size = 0;

    constructor(items: (CacheItem<K, V> | SerializableCacheItem<K, V>)[]) {
        const now = Date.now();

        for (let i = items.length; i > 0; i--) {
            const item = items[i - 1];
            const cacheItem: CacheItem<K, V> =
                item instanceof CacheItem ? item : new CacheItem<K, V>(item.key, item.value);
            cacheItem.isValid = true;
            cacheItem.lastUpdateAt = now;
            this.prepend(cacheItem);
        }
    }

    prepend(item: CacheItem<K, V>): void {
        item[NEXT] = this.head;
        if (this.head) {
            this.head[PREV] = item;
        }
        this.head = item;
        if (!this.tail) {
            this.tail = item;
        }
        this.size += 1;
    }

    remove(item: CacheItem<K, V>): boolean {
        if (!item.key) {
            return false;
        }

        const next = item[NEXT];
        const prev = item[PREV];

        if (prev) {
            prev[NEXT] = next;
        } else {
            this.head = next;
        }

        if (next) {
            next[PREV] = prev;
        } else {
            this.tail = prev;
        }

        this.size -= 1;

        return true;
    }

    findBy(predicate: (item: CacheItem<K, V>) => boolean): CacheItem<K, V> | undefined {
        return reduce<CacheItem<K, V> | undefined, K, V>(this, undefined, (acc, item) =>
            predicate(item) ? reduced(item) : acc,
        );
    }
}
