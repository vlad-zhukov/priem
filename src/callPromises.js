import moize from 'moize';
import {elementsEqual} from 'react-shallow-equal';
import * as promiseState from './promiseState';
import {type} from './helpers';

const memoized = {};

function memoizeAsyncValue(key, value, onExpire) {
    const typeOfValue = type(value);
    if (typeOfValue !== 'object') {
        throw new TypeError(`Priem: argument 'values' must return an object of objects, but got: '${typeOfValue}'.`);
    }

    const typeOfValuePromise = type(value.promise);
    if (typeOfValuePromise !== 'function') {
        throw new TypeError(
            "Priem: argument 'values' must return an object of objects with a property 'promise' " +
                `each, but got: '${typeOfValuePromise}'.`
        );
    }

    memoized[key] = moize(value.promise, {
        isPromise: true,
        maxAge: value.maxAge,
        maxArgs: value.maxArgs,
        maxSize: value.maxSize,
        onExpire,
    });
}

export function extractAsyncValues(props) {
    if (type(props.asyncValues) !== 'function') {
        return {};
    }

    const values = props.asyncValues(props);

    const typeOfValues = type(values);
    if (typeOfValues !== 'object') {
        throw new TypeError(
            `Priem: property 'asyncValues' must be a function that returns an object, but got: '${typeOfValues}'.`
        );
    }

    return values;
}

function callPromise({props, key, asyncValue, prevArgs, onExpire, isMounting, isForced}) {
    // Do not recall rejected (uncached) promises unless forced
    if (
        isForced === false &&
        props.priem !== undefined &&
        props.priem[key] !== undefined &&
        props.priem[key].rejected === true
    ) {
        return;
    }

    const args = type(asyncValue.args) === 'array' ? asyncValue.args : [];

    if (memoized[key] !== undefined) {
        if (memoized[key].has(args) === true) {
            if (isMounting === false && elementsEqual(args, prevArgs)) {
                return;
            }
        }
        else {
            // Only set to refreshing when the result is not cached
            props.setPriem(s => ({
                [key]: promiseState.refreshing(s[key]),
            }));
        }
    }
    else {
        memoizeAsyncValue(key, asyncValue, onExpire);

        if (isMounting === false) {
            props.setPriem({
                [key]: promiseState.pending(),
            });
        }
    }

    memoized[key](...args)
        .then((result) => {
            props.setPriem({
                [key]: promiseState.fulfilled(result),
            });
        })
        .catch((e) => {
            props.setPriem({
                [key]: promiseState.rejected(e.message),
            });
        });
}

export function callPromises({props, prevProps, onExpire, isMounting = false, isForced = false}) {
    // Stop auto-refreshing
    if (props.autoRefresh === false && isForced === false) {
        return;
    }

    const asyncValues = extractAsyncValues(props);
    const prevAsyncValues = prevProps && prevProps.priem ? extractAsyncValues(prevProps) : null;
    const asyncKeys = Object.keys(asyncValues);
    for (let i = 0, l = asyncKeys.length; i < l; i++) {
        const key = asyncKeys[i];
        const asyncValue = asyncValues[key];

        const prevArgs =
            prevAsyncValues && type(prevAsyncValues[key].args) === 'array' ? prevAsyncValues[key].args : [];

        callPromise({props, key, asyncValue, prevArgs, onExpire, isMounting, isForced});
    }
}
