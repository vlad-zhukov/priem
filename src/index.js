export {default as Priem} from './Priem';
export {Container, populateStore, flushStore} from './Container';
export {default as withPriem} from './withPriem';
export {default as getDataFromTree} from './getDataFromTree';
