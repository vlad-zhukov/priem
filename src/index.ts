export {createResource, GetArgs, Options, Result, ResultMeta, ResultPages, ResultPagesMeta} from './createResource';
export {
    hydrateStore,
    flushStore,
    getRunningPromises,
    ResourceOptions,
    MemoizedKey,
    MemoizedValue,
    MemoizedSerializableCacheItem,
    Status,
} from './Resource';
export {SerializableCacheItem} from './Cache';
