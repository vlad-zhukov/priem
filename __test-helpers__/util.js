/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import delay from 'delay';
import Priem from '../src/Priem';
import {Container, populateStore} from '../src/Container';

export function testComponent({initialStore, options} = {}) {
    if (initialStore) {
        populateStore(initialStore);
    }

    const ctr = new Container({
        mapPropsToArgs: () => ['foo'],
        promise: value => delay(100, {value}),
        ...options,
    });

    return <Priem sources={{ctr}}>{p => <div>{p.ctr}</div>}</Priem>;
}

export function testComponentNested({initialStore, ctr1Props, ctr2Props} = {}) {
    if (initialStore) {
        populateStore(initialStore);
    }

    const ctr1 = new Container({
        mapPropsToArgs: () => ['foo'],
        promise: value => delay(100, {value}),
        ...ctr1Props,
    });

    const ctr2 = new Container({
        mapPropsToArgs: p => (!p.ctr1 ? null : [p.ctr1, 'bar']),
        promise: (ctr1Value, value) => delay(100, {value: ctr1Value + value}),
        ...ctr2Props,
    });

    return (
        <Priem sources={{ctr1, ctr2}}>
            {props => (
                <Priem sources={{ctr2}} ctr1={props.ctr1}>
                    {p => <div>{p.ctr2}</div>}
                </Priem>
            )}
        </Priem>
    );
}

/* eslint-disable react/no-unused-state */
export class ErrorBoundary extends React.Component {
    constructor() {
        super();
        this.state = {initTime: Date.now(), hasError: null};
    }

    componentDidCatch(error) {
        this.setState({hasError: error, catchTime: Date.now()});
    }

    render() {
        if (this.state.hasError) {
            return null;
        }
        return this.props.children;
    }
}
/* eslint-enable react/no-unused-state */
