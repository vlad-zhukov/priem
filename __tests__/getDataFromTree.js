/**
 * @jest-environment node
 */

import React from 'react';
import ReactDOM from 'react-dom/server';
import {getDataFromTree} from '../src/index';
import {TestComponentSimple, TestComponentNested} from '../__test-helpers__/util';

it('should fetch and render to string with data', async () => {
    const data = await getDataFromTree(<TestComponentSimple />);
    expect(data).toMatchSnapshot();

    const content = ReactDOM.renderToStaticMarkup(<TestComponentSimple initialState={data} />);
    expect(content).toBe('<div>foo</div>');
});

it('should not fetch data from a nested component', async () => {
    const data = await getDataFromTree(<TestComponentNested />);
    expect(data).toMatchSnapshot();

    const content = ReactDOM.renderToStaticMarkup(<TestComponentNested initialState={data} />);
    expect(content).toBe('');
});
