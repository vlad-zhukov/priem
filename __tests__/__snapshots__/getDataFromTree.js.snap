// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fetch and render to string with data 1`] = `
Array [
  Object {
    "meta": Object {
      "args": Array [
        "foo",
      ],
      "ssr": true,
    },
    "state": Object {
      "fulfilled": true,
      "pending": false,
      "reason": null,
      "refreshing": false,
      "rejected": false,
      "value": "foo",
    },
  },
]
`;

exports[`should not fetch data from a nested component 1`] = `
Array [
  Object {
    "meta": undefined,
    "state": Object {
      "counter": 2,
    },
  },
  Object {
    "meta": Object {
      "args": Array [
        2,
        "foo",
      ],
      "ssr": true,
    },
    "state": Object {
      "fulfilled": true,
      "pending": false,
      "reason": null,
      "refreshing": false,
      "rejected": false,
      "value": "2-foo",
    },
  },
]
`;
